# <span style = 'color: #3D9970'> **00. Introdução:** </span>

# <span style = 'color: #3D9970'> **01. Carregar dados e módulos:** </span>

## <span style = 'color: #39CCCC'> **Carregando módulos e bibliotecas:** </span>

In [267]:
import pandas as pd # manipulação e análise de dados
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import locale

from sklearn.impute import SimpleImputer

#### <span style = 'color: FFFFFF'> **Configurando o locale para *pt-br*:** </span>

In [268]:
locale.setlocale(locale.LC_ALL, 'pt_BR.utf8')

'pt_BR.utf8'

## <span style = 'color: #39CCCC'> **Instalando, ativando e desativando maquina virtual:** </span>

In [269]:
# Para instalar um ambiente virtual, abra o terminal e digite:
# ------------------------------------------------------------ 
# sudo pip3 install virtualenv
# ------------------------------------------------------------ 
# Para criar um ambiente virtual, abra o terminal no diretório desejado e digite:
# ------------------------------------------------------------ 
# virtualenv venv
# ------------------------------------------------------------ 
# Para inicar um ambiente virtual, abra o terminal no diretório desejado e digite:
# ------------------------------------------------------------ 
# source venv/bin/activate
# ------------------------------------------------------------ 
# Para desativar um ambiente virtual, abra o terminal no diretório desejado e digite:
# ------------------------------------------------------------ 
# deactivate
# ------------------------------------------------------------ 

## <span style = 'color: #39CCCC'> **Carregando dataset's:** </span>

In [270]:
olist_customers_dataset = pd.read_csv('../data/raw/olist_customers_dataset.csv') 
olist_geolocation_dataset = pd.read_csv('../data/raw/olist_geolocation_dataset.csv')
olist_order_items_dataset = pd.read_csv('../data/raw/olist_order_items_dataset.csv')
olist_order_payments_dataset = pd.read_csv('../data/raw/olist_order_payments_dataset.csv')
olist_order_reviews_dataset = pd.read_csv('../data/raw/olist_order_reviews_dataset.csv')
olist_orders_dataset = pd.read_csv('../data/raw/olist_orders_dataset.csv')
olist_products_dataset = pd.read_csv('../data/raw/olist_products_dataset.csv')
olist_sellers_dataset = pd.read_csv('../data/raw/olist_sellers_dataset.csv')
product_category_name_translation = pd.read_csv('../data/raw/product_category_name_translation.csv')

## <span style = 'color: #39CCCC'> **Funções:** </span>

In [271]:
def visualizar_tabela(dataset):
    return dataset.head(5)

def valores_nulos_dataset(dataset):
    print(dataset.isnull().sum())
    print(dataset.info())

def info_descritiva(dataset):
    return dataset.describe()

def colunas_dataset(dataset):
    return list(dataset.columns.values)

def valores_colunas(dataset, coluna):
    return dataset[coluna].unique()

def formatar_valores_para_reais(dataset, coluna1, coluna2):
    dataset[coluna1] = 'R$ ' + dataset[coluna2].round(2).astype(str)
    return dataset

# <span style = 'color: #3D9970'> **02. Exploração inicial:** </span>

## <span style = 'color: #39CCCC'> **Conhecendo os Datasets's:** </span>

### <span style = 'color: #0074D9'> **Clientes:** </span>

In [272]:
visualizar_tabela(olist_customers_dataset)

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP  
3                      8775        mogi das cruzes             SP  
4                     13056               campinas             SP

In [273]:
info_descritiva(olist_customers_dataset) 

customer_zip_code_prefix
count              99441.000000
mean               35137.474583
std                29797.938996
min                 1003.000000
25%                11347.000000
50%                24416.000000
75%                58900.000000
max                99990.000000

In [274]:
valores_nulos_dataset(olist_customers_dataset)

customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   customer_id               99441 non-null  object
 1   customer_unique_id        99441 non-null  object
 2   customer_zip_code_prefix  99441 non-null  int64 
 3   customer_city             99441 non-null  object
 4   customer_state            99441 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.8+ MB
None


#### <span style = 'color: #cc9900'> *Análise das colunas:* </span>

| Colunas                  | Explicação                                               |
|--------------------------|----------------------------------------------------------|
| customer_id              | Esta coluna nos informa o ID de cada cliente.            |
| customer_unique_id       | Esta coluna nos informa o ID especifico de cada cliente. |
| customer_zip_code_prefix | Esta coluna nos informa o CEP de cada cliente.           |
| customer_city            | Esta coluna nos informa a cidade de cada cliente.        |
| customer_state           | Esta coluna nos informa o estado de cada cliente.        |

#### <span style = 'color: #cc9900'> *Comentários sobre o dataset:* </span>

- O dataset possui 5 colunas, sendo a única que não é um do tipo "object" é a coluna "customer_zip_code_prefix", sendo a mesma, do tipo "int64".
- O dataset não possui valor nulo em nenhuma de suas colunas.
- Para este dataset, não é necessário investigar dados estatistícos descritivos, já que a coluna que é do tipo "int", é uma coluna qualitativa.
- Por fim, podemos concluir que, este dataset, nos entrega dados a respeito do cliente de cada venda.

### <span style = 'color: #0074D9'> **Geolocalização:** </span>

In [275]:
visualizar_tabela(olist_geolocation_dataset)

geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
0                         1037       -23.545621       -46.639292   
1                         1046       -23.546081       -46.644820   
2                         1046       -23.546129       -46.642951   
3                         1041       -23.544392       -46.639499   
4                         1035       -23.541578       -46.641607   

  geolocation_city geolocation_state  
0        sao paulo                SP  
1        sao paulo                SP  
2        sao paulo                SP  
3        sao paulo                SP  
4        sao paulo                SP

In [276]:
info_descritiva(olist_geolocation_dataset) 

geolocation_zip_code_prefix  geolocation_lat  geolocation_lng
count                 1.000163e+06     1.000163e+06     1.000163e+06
mean                  3.657417e+04    -2.117615e+01    -4.639054e+01
std                   3.054934e+04     5.715866e+00     4.269748e+00
min                   1.001000e+03    -3.660537e+01    -1.014668e+02
25%                   1.107500e+04    -2.360355e+01    -4.857317e+01
50%                   2.653000e+04    -2.291938e+01    -4.663788e+01
75%                   6.350400e+04    -1.997962e+01    -4.376771e+01
max                   9.999000e+04     4.506593e+01     1.211054e+02

In [277]:
valores_nulos_dataset(olist_geolocation_dataset)

geolocation_zip_code_prefix    0
geolocation_lat                0
geolocation_lng                0
geolocation_city               0
geolocation_state              0
dtype: int64
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000163 entries, 0 to 1000162
Data columns (total 5 columns):
 #   Column                       Non-Null Count    Dtype  
---  ------                       --------------    -----  
 0   geolocation_zip_code_prefix  1000163 non-null  int64  
 1   geolocation_lat              1000163 non-null  float64
 2   geolocation_lng              1000163 non-null  float64
 3   geolocation_city             1000163 non-null  object 
 4   geolocation_state            1000163 non-null  object 
dtypes: float64(2), int64(1), object(2)
memory usage: 38.2+ MB
None


#### <span style = 'color: #cc9900'> *Análise das colunas:* </span>

| Colunas                     | Explicação                                             |
|-----------------------------|--------------------------------------------------------|
| geolocation_zip_code_prefix | Esta coluna nos informa o CEP.                         |
| geolocation_lat             | Esta coluna nos informa a posição de latitude do CEP.  |
| geolocation_lng             | Esta coluna nos informa a posição de longitude do CEP. |
| geolocation_city            | Esta coluna nos informa a cidade do CEP.               |
| geolocation_state           | Esta coluna nos informa o estado do CEP.               |

#### <span style = 'color: #cc9900'> *Comentários sobre o dataset:* </span>

- O dataset possui 5 colunas, sendo 1 coluna do tipo "int64" ("geolocation_zip_code_prefix"), 2 do tipo "float64" ("geolocation_lat" e "geolocation_lng") e por fim, 2 do tipo "object" ("geolocation_city" e "geolocation_state").
- O dataset não possui valor nulo em nenhuma de suas colunas.
- Para este dataset, não é necessário investigar dados estatistícos descritivos, já que as colunas que são do tipo "int64" e "float64", são colunas qualitativas.
- Por fim, podemos concluir que, este dataset, nos entrega dados do CEP.

### <span style = 'color: #0074D9'> **Pedido de compra:** </span>

In [278]:
visualizar_tabela(olist_order_items_dataset)

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

   shipping_limit_date   price  freight_value  
0  2017-09-19 09:45:35   58.90          13.29  
1  2017-05-03 11:05:13  239.90          19.93  
2  2018-01-18 14:48:30  199.00          17.87  
3  2018-08-15 10:10:18   12.99          12.79  
4  2017-02-13 13:57:51  199.90          18.14

In [279]:
info_descritiva(olist_order_items_dataset) 

order_item_id          price  freight_value
count  112650.000000  112650.000000  112650.000000
mean        1.197834     120.653739      19.990320
std         0.705124     183.633928      15.806405
min         1.000000       0.850000       0.000000
25%         1.000000      39.900000      13.080000
50%         1.000000      74.990000      16.260000
75%         1.000000     134.900000      21.150000
max        21.000000    6735.000000     409.680000

In [280]:
valores_nulos_dataset(olist_order_items_dataset)

order_id               0
order_item_id          0
product_id             0
seller_id              0
shipping_limit_date    0
price                  0
freight_value          0
dtype: int64
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   order_id             112650 non-null  object 
 1   order_item_id        112650 non-null  int64  
 2   product_id           112650 non-null  object 
 3   seller_id            112650 non-null  object 
 4   shipping_limit_date  112650 non-null  object 
 5   price                112650 non-null  float64
 6   freight_value        112650 non-null  float64
dtypes: float64(2), int64(1), object(4)
memory usage: 6.0+ MB
None


#### <span style = 'color: #cc9900'> *Análise das colunas:* </span>

| Colunas             | Explicação                                                       |
|---------------------|------------------------------------------------------------------|
| order_id            | Esta coluna representa o ID do pedido.                           |
| order_item_id       | Esta coluna representa a quantidade de produtos no pedido.       |
| product_id          | Esta coluna representa o ID do produto no pedido.                |
| seller_id           | Esta coluna representa o ID da venda.                            |
| shipping_limit_date | Esta coluna representa ao prazo máximo de "despacho" do pedido.  |
| price               | Esta coluna representa os valores de cada pedido.                |
| freight_value       | Esta coluna representa os valores pagos no frete de cada pedido. |

#### <span style = 'color: #cc9900'> *Comentários sobre o dataset:* </span>

- O dataset possui 7 colunas, sendo 4 colunas do tipo "object" ("order_id", "product_id", "seller_id" e "shipping_limit_date"), 2 do tipo "float64" ("price" e "freight_value") e por fim, 1 coluna do tipo "int64" (order_item_id).
- O dataset não possui valor nulo em nenhuma de suas colunas.
- Realizando análises estatísticas descritivas neste dataset, podemos observar que há 112.650 pedidos, sendo que desses pedidos, a grande maioria possui uma média de 1 produto, a média de preço é de R$ 120,00 e a média de frete é de R$ 20,00. Além disso, o produto pedido mais caro possui o valor de R$ 6.735,00 e o maior frete de R$ 409,00.
- Por fim, podemos concluir que este dataset nos da informações a respeito dos pedidos realizados.

### <span style = 'color: #0074D9'> **Pagamentos:** </span>

In [281]:
visualizar_tabela(olist_order_payments_dataset)

order_id  payment_sequential payment_type  \
0  b81ef226f3fe1789b1e8b2acac839d17                   1  credit_card   
1  a9810da82917af2d9aefd1278f1dcfa0                   1  credit_card   
2  25e8ea4e93396b6fa0d3dd708e76c1bd                   1  credit_card   
3  ba78997921bbcdc1373bb41e913ab953                   1  credit_card   
4  42fdf880ba16b47b59251dd489d4441a                   1  credit_card   

   payment_installments  payment_value  
0                     8          99.33  
1                     1          24.39  
2                     1          65.71  
3                     8         107.78  
4                     2         128.45

In [282]:
info_descritiva(olist_order_payments_dataset) 

payment_sequential  payment_installments  payment_value
count       103886.000000         103886.000000  103886.000000
mean             1.092679              2.853349     154.100380
std              0.706584              2.687051     217.494064
min              1.000000              0.000000       0.000000
25%              1.000000              1.000000      56.790000
50%              1.000000              1.000000     100.000000
75%              1.000000              4.000000     171.837500
max             29.000000             24.000000   13664.080000

In [283]:
valores_nulos_dataset(olist_order_payments_dataset)

order_id                0
payment_sequential      0
payment_type            0
payment_installments    0
payment_value           0
dtype: int64
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103886 entries, 0 to 103885
Data columns (total 5 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   order_id              103886 non-null  object 
 1   payment_sequential    103886 non-null  int64  
 2   payment_type          103886 non-null  object 
 3   payment_installments  103886 non-null  int64  
 4   payment_value         103886 non-null  float64
dtypes: float64(1), int64(2), object(2)
memory usage: 4.0+ MB
None


#### <span style = 'color: #cc9900'> *Análise das colunas:* </span>

| Colunas              | Explicação                                                                                           |
|----------------------|------------------------------------------------------------------------------------------------------|
| order_id             | Esta coluna representa o ID do pedido.                                                               |
| payment_sequential   | Esta coluna representa a quantidade de métodos de pagamentos utilizadas no pedido.                   |
| payment_type         | Esta coluna representa o tipo de pagamento realizado.                                                |
| payment_installments | Esta coluna representa a quantidade de parcelas.                                                     |
| payment_value        | Esta coluna representa o valor pago.                                                                 |

#### <span style = 'color: #cc9900'> *Comentários sobre o dataset:* </span>

- O dataset possui 5 colunas, sendo 2 do tipo "object" ("order_id" e "payment_type"), 2 do tipo "int64" ("payment_sequential" e "payment_installments") e por fim, 1 coluna do tipo "float64" ("payment_value").
- O dataset não possui valor nulo em nenhuma de suas colunas.
- Realizando análises estatísticas descritivas neste dataset, podemos observar que há 103.886 pedidos, sendo que desses pedidos, a valor médio pago por cada pedido é de R$ 154,00 e o máximo de R$ 13.664,00. Além disso, vemos que há a média de parcelas por pedido é de 2 parcelas e o máximo de parcelas realizadas foi de 24.
- Por fim, podemos concluir que este dataset nos da informações a respeito dos pagamentos de cada pedido.

### <span style = 'color: #0074D9'> **Feedback's:** </span>

In [284]:
visualizar_tabela(olist_order_reviews_dataset)

review_id                          order_id  \
0  7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1  80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2  228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3  e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4  f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   

   review_score review_comment_title  \
0             4                  NaN   
1             5                  NaN   
2             5                  NaN   
3             5                  NaN   
4             5                  NaN   

                              review_comment_message review_creation_date  \
0                                                NaN  2018-01-18 00:00:00   
1                                                NaN  2018-03-10 00:00:00   
2                                                NaN  2018-02-17 00:00:00   
3              Recebi bem antes do prazo estipulado.  2017-04-21 00:00:00   
4  Parabéns lojas lannister adorei comprar pela I...  2018-03-01 00:00:00   

  review_answer_timestamp  
0     2018-01-18 21:46:59  
1     2018-03-11 03:05:13  
2     2018-02-18 14:36:24  
3     2017-04-21 22:02:06  
4     2018-03-02 10:26:53

In [285]:
info_descritiva(olist_order_reviews_dataset)

review_score
count  99224.000000
mean       4.086421
std        1.347579
min        1.000000
25%        4.000000
50%        5.000000
75%        5.000000
max        5.000000

In [286]:
valores_nulos_dataset(olist_order_reviews_dataset)

review_id                      0
order_id                       0
review_score                   0
review_comment_title       87656
review_comment_message     58247
review_creation_date           0
review_answer_timestamp        0
dtype: int64
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99224 entries, 0 to 99223
Data columns (total 7 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   review_id                99224 non-null  object
 1   order_id                 99224 non-null  object
 2   review_score             99224 non-null  int64 
 3   review_comment_title     11568 non-null  object
 4   review_comment_message   40977 non-null  object
 5   review_creation_date     99224 non-null  object
 6   review_answer_timestamp  99224 non-null  object
dtypes: int64(1), object(6)
memory usage: 5.3+ MB
None


#### <span style = 'color: #cc9900'> *Análise das colunas:* </span>

| Colunas                 | Explicação                                                     |
|-------------------------|----------------------------------------------------------------|
| review_id               | Esta coluna representa o ID do review.                         |
| order_id                | Esta coluna representa o ID do pedido.                         |
| review_score            | Esta coluna representa a pontuação dada ao pedido (1-5).       |
| review_comment_title    | Esta coluna representa o titulo do comentário do review.       |
| review_comment_message  | Esta coluna representa o comentário do review.                 |
| review_creation_date    | Esta coluna representa a data que o comentário foi criado.     |
| review_answer_timestamp | Esta coluna representa a data que o comentário foi respondido. |

#### <span style = 'color: #cc9900'> *Comentários sobre o dataset:* </span>

- O dataset possui 7 colunas, sendo a única que não é um do tipo "object" é a coluna "review_score", sendo a mesma, do tipo "int64".
- De todas as colunas, as que possuem valores nulos são as "review_comment_title" e "review_comment_message".
- Podemos concluir que todas os reviews foram respondidas, pois a coluna "review_answer_timestamp" não possui valor nulo.
- Há uma boa média com relação a avaliação dos produtos, pois na coluna "review_score" podemos observar que a média da coluna é de 4.0 e possui um desvio padrão de 1.3.
- Por fim, podemos investigar o motivo de algumas notas possuirem um valor de "review_score" baixo.

### <span style = 'color: #0074D9'> **Entrega do pedido:** </span>

In [287]:
visualizar_tabela(olist_orders_dataset)

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  
0           2017-10-18 00:00:00  
1           2018-08-13 00:00:00  
2           2018-09-04 00:00:00  
3           2017-12-15 00:00:00  
4           2018-02-26 00:00:00

In [288]:
info_descritiva(olist_orders_dataset)

order_id                       customer_id  \
count                              99441                             99441   
unique                             99441                             99441   
top     e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
freq                                   1                                 1   

       order_status order_purchase_timestamp    order_approved_at  \
count         99441                    99441                99281   
unique            8                    98875                90733   
top       delivered      2018-04-11 10:48:14  2018-02-27 04:31:10   
freq          96478                        3                    9   

       order_delivered_carrier_date order_delivered_customer_date  \
count                         97658                         96476   
unique                        81018                         95664   
top             2018-05-09 15:48:00           2018-05-08 23:38:46   
freq                             47                             3   

       order_estimated_delivery_date  
count                          99441  
unique                           459  
top              2017-12-20 00:00:00  
freq                             522

In [289]:
valores_nulos_dataset(olist_orders_dataset)

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   order_id                       99441 non-null  object
 1   customer_id                    99441 non-null  object
 2   order_status                   99441 non-null  object
 3   order_purchase_timestamp       99441 non-null  object
 4   order_approved_at              99281 non-null  object
 5   order_delivered_carrier_date   97658 non-null  object
 6   order_delivered_customer_date  96476 non-null  object
 7   order_estimated_delivery_date  99441 no

#### <span style = 'color: #cc9900'> *Análise das colunas:* </span>

| Colunas                       | Explicação                                                                  |
|-------------------------------|-----------------------------------------------------------------------------|
| order_id                      | Esta coluna representa o ID do pedido.                                      |
| customer_id                   | Esta coluna representa o ID do cliente.                                     |
| order_status                  | Esta coluna representa o status do pedido com relação a entrega.            |
| order_purchase_timestamp      | Esta coluna representa a data de pagamento.                                 |
| order_approved_at             | Esta coluna representa a aprovação de pagamento do pedido.                  |
| order_delivered_carrier_date  | Esta coluna representa a data que o pedido foi entregue a transportadora.   |
| order_delivered_customer_date | Esta coluna representa a data que o cliente recebeu o pedido.               |
| order_estimated_delivery_date | Esta coluna representa a data estimada que o cliente iria receber o pedido. |

#### <span style = 'color: #cc9900'> *Comentários sobre o dataset:* </span>

- O dataset possui 7 colunas, sendo que todas estão no tipo "object"
- Existem colunas com dados nulos como "order_approved_at", "order_delivered_carrier_date" e "order_delivered_customer_date", podendo ser possíveis motivos para avaliações de produtos baixas.
- Podemos concluir que este é um dataset que representa o status e os dados a respeito do pagamento de seu pedido até a entrega do mesmo. 

### <span style = 'color: #0074D9'> **Produtos:** </span>

In [290]:
visualizar_tabela(olist_products_dataset)

product_id  product_category_name  \
0  1e9e8ef04dbcff4541ed26657ea517e5             perfumaria   
1  3aa071139cb16b67ca9e5dea641aaa2f                  artes   
2  96bd76ec8810374ed1b65e291975717f          esporte_lazer   
3  cef67bcfe19066a932b7673e239eb23d                  bebes   
4  9dc1a7de274444849c219cff195d0b71  utilidades_domesticas   

   product_name_lenght  product_description_lenght  product_photos_qty  \
0                 40.0                       287.0                 1.0   
1                 44.0                       276.0                 1.0   
2                 46.0                       250.0                 1.0   
3                 27.0                       261.0                 1.0   
4                 37.0                       402.0                 4.0   

   product_weight_g  product_length_cm  product_height_cm  product_width_cm  
0             225.0               16.0               10.0              14.0  
1            1000.0               30.0               18.0              20.0  
2             154.0               18.0                9.0              15.0  
3             371.0               26.0                4.0              26.0  
4             625.0               20.0               17.0              13.0

In [291]:
info_descritiva(olist_products_dataset)

product_name_lenght  product_description_lenght  product_photos_qty  \
count         32341.000000                32341.000000        32341.000000   
mean             48.476949                  771.495285            2.188986   
std              10.245741                  635.115225            1.736766   
min               5.000000                    4.000000            1.000000   
25%              42.000000                  339.000000            1.000000   
50%              51.000000                  595.000000            1.000000   
75%              57.000000                  972.000000            3.000000   
max              76.000000                 3992.000000           20.000000   

       product_weight_g  product_length_cm  product_height_cm  \
count      32949.000000       32949.000000       32949.000000   
mean        2276.472488          30.815078          16.937661   
std         4282.038731          16.914458          13.637554   
min            0.000000           7.000000           2.000000   
25%          300.000000          18.000000           8.000000   
50%          700.000000          25.000000          13.000000   
75%         1900.000000          38.000000          21.000000   
max        40425.000000         105.000000         105.000000   

       product_width_cm  
count      32949.000000  
mean          23.196728  
std           12.079047  
min            6.000000  
25%           15.000000  
50%           20.000000  
75%           30.000000  
max          118.000000

In [292]:
valores_nulos_dataset(olist_products_dataset)

product_id                      0
product_category_name         610
product_name_lenght           610
product_description_lenght    610
product_photos_qty            610
product_weight_g                2
product_length_cm               2
product_height_cm               2
product_width_cm                2
dtype: int64
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32951 non-null  object 
 1   product_category_name       32341 non-null  object 
 2   product_name_lenght         32341 non-null  float64
 3   product_description_lenght  32341 non-null  float64
 4   product_photos_qty          32341 non-null  float64
 5   product_weight_g            32949 non-null  float64
 6   product_length_cm           32949 non-null  float64
 7   product_height_cm           32949 non-null  float64
 8 

#### <span style = 'color: #cc9900'> *Análise das colunas:* </span>

| Colunas                    | Explicação                                                                         |
|----------------------------|------------------------------------------------------------------------------------|
| product_id                 | Esta coluna representa o ID do produto.                                            |
| product_category_name      | Esta coluna representa a categoria do produto.                                     |
| product_name_lenght        | Esta coluna representa a quantidade de caracter que possui o nome do produto.      |
| product_description_lenght | Esta coluna representa a quantidade de caracter que possui a descrição do produto. |
| product_photos_qty         | Esta coluna representa a quantidade de fotos que o produto possui.                 |
| product_weight_g           | Esta coluna representa o peso do produto em gramas (g).                            |
| product_length_cm          | Esta coluna representa o comprimento do produto em centimetros (cm).               |
| product_height_cm          | Esta coluna representa a altura do produto em centimetros (cm).                    |
| product_width_cm           | Esta coluna representa a largura do produto em centimetros (cm).                   |

#### <span style = 'color: #cc9900'> *Comentários sobre o dataset:* </span>

- O dataset possui 9 colunas, sendo que destas colunas, 2 são do tipo "object" ("product_id" e "product_category_name") e as outras são do tipo "float64".
- Existem colunas com dados nulos como "product_category_name", "product_name_lenght", "product_description_lenght", "product_photos_qty", "product_weight_g", "product_length_cm", "product_height_cm" e "product_width_cm". Essas variáveis nos indicam que nem todos os produtos possuem todas suas descrições.
- Algumas análises que podemos verificar é que a média de imagem por produto é de 2 fotos, que a média de massa dos produtos é de 2 kg e o produto com maior massa é de 40 kg. 
- Podemos concluir que este dataset nos da informações a respeito do produto, como categoria, peso e afins.

### <span style = 'color: #0074D9'> **Vendas:** </span>

In [293]:
visualizar_tabela(olist_sellers_dataset)

seller_id  seller_zip_code_prefix  \
0  3442f8959a84dea7ee197c632cb2df15                   13023   
1  d1b65fc7debc3361ea86b5f14c68d2e2                   13844   
2  ce3ad9de960102d0677a81f5d0bb7b2d                   20031   
3  c0f3eea2e14555b6faeea3dd58c1b1c3                    4195   
4  51a04a8a6bdcb23deccc82b0b80742cf                   12914   

         seller_city seller_state  
0           campinas           SP  
1         mogi guacu           SP  
2     rio de janeiro           RJ  
3          sao paulo           SP  
4  braganca paulista           SP

In [294]:
info_descritiva(olist_sellers_dataset)

seller_zip_code_prefix
count             3095.000000
mean             32291.059451
std              32713.453830
min               1001.000000
25%               7093.500000
50%              14940.000000
75%              64552.500000
max              99730.000000

In [295]:
valores_nulos_dataset(olist_sellers_dataset)

seller_id                 0
seller_zip_code_prefix    0
seller_city               0
seller_state              0
dtype: int64
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3095 entries, 0 to 3094
Data columns (total 4 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   seller_id               3095 non-null   object
 1   seller_zip_code_prefix  3095 non-null   int64 
 2   seller_city             3095 non-null   object
 3   seller_state            3095 non-null   object
dtypes: int64(1), object(3)
memory usage: 96.8+ KB
None


#### <span style = 'color: #cc9900'> *Análise das colunas:* </span>

| Colunas                | Explicação                                                                 |
|------------------------|----------------------------------------------------------------------------|
| seller_id              | Esta coluna representa o ID da venda.                                      |
| seller_zip_code_prefix | Esta coluna representa o CEP do endereço da venda.                         |
| seller_city            | Esta coluna representa a cidade do endereço da venda.                      |
| seller_state           | Esta coluna representa o estado do endereço da venda.                      | 

#### <span style = 'color: #cc9900'> *Comentários sobre o dataset:* </span>

- O dataset possui 4 colunas, sendo que destas colunas, 3 são do tipo "object" ("seller_id", "seller_city" e "seller_state") e 1 é do tipo "seller_zip_code_prefix".
- Existem dataset não possui valor nulo.
- Não podemos realizar análises descritivas, pois os dados que estão presentes no dataset são dados qualitativos.
- Podemos concluir que este dataset nos dos endereços das vendas.

### <span style = 'color: #0074D9'> **Tradução de categorias (para o inglês):** </span>

In [296]:
visualizar_tabela(product_category_name_translation)

product_category_name product_category_name_english
0            beleza_saude                 health_beauty
1  informatica_acessorios         computers_accessories
2              automotivo                          auto
3         cama_mesa_banho                bed_bath_table
4        moveis_decoracao               furniture_decor

In [297]:
valores_nulos_dataset(product_category_name_translation)

product_category_name            0
product_category_name_english    0
dtype: int64
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 71 entries, 0 to 70
Data columns (total 2 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   product_category_name          71 non-null     object
 1   product_category_name_english  71 non-null     object
dtypes: object(2)
memory usage: 1.2+ KB
None


#### <span style = 'color: #cc9900'> *Comentários sobre o dataset:* </span>

- Este dataset não será utilizado, já que sua única utilidade seria para traduzir o nome da categoria dos produtos para o inglês.

## <span style = 'color: #39CCCC'> **Analisando o dataset "olist_order_reviews_dataset":** </span>

### <span style = 'color: #0074D9'> **Motivos da análise:** </span>

- Entender quais variáveis levam um feedback ter nota baixa.
- Para realizar está investigação, devemos entender melhor o dataset "olist_order_reviews_dataset" e sua coluna "review_score".

### <span style = 'color: #0074D9'> **Dataset:** </span>


In [298]:
colunas_dataset(olist_order_reviews_dataset)

['review_id',
 'order_id',
 'review_score',
 'review_comment_title',
 'review_comment_message',
 'review_creation_date',
 'review_answer_timestamp']

In [299]:
visualizar_tabela(olist_order_reviews_dataset)

review_id                          order_id  \
0  7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1  80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2  228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3  e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4  f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   

   review_score review_comment_title  \
0             4                  NaN   
1             5                  NaN   
2             5                  NaN   
3             5                  NaN   
4             5                  NaN   

                              review_comment_message review_creation_date  \
0                                                NaN  2018-01-18 00:00:00   
1                                                NaN  2018-03-10 00:00:00   
2                                                NaN  2018-02-17 00:00:00   
3              Recebi bem antes do prazo estipulado.  2017-04-21 00:00:00   
4  Parabéns lojas lannister adorei comprar pela I...  2018-03-01 00:00:00   

  review_answer_timestamp  
0     2018-01-18 21:46:59  
1     2018-03-11 03:05:13  
2     2018-02-18 14:36:24  
3     2017-04-21 22:02:06  
4     2018-03-02 10:26:53

In [300]:
olist_order_reviews_dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99224 entries, 0 to 99223
Data columns (total 7 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   review_id                99224 non-null  object
 1   order_id                 99224 non-null  object
 2   review_score             99224 non-null  int64 
 3   review_comment_title     11568 non-null  object
 4   review_comment_message   40977 non-null  object
 5   review_creation_date     99224 non-null  object
 6   review_answer_timestamp  99224 non-null  object
dtypes: int64(1), object(6)
memory usage: 5.3+ MB


### <span style = 'color: #0074D9'> **Visualizando os valores das variáveis:** </span>

In [301]:
valores_colunas(olist_order_reviews_dataset, 'review_id')

array(['7bc2406110b926393aa56f80a40eba40',
       '80e641a11e56f04c1ad469d5645fdfde',
       '228ce5500dc1d8e020d8d1322874b6f0', ...,
       'b3de70c89b1510c4cd3d0649fd302472',
       '1adeb9d84d72fe4e337617733eb85149',
       'efe49f1d6f951dd88b51e6ccd4cc548f'], dtype=object)

In [302]:
valores_colunas(olist_order_reviews_dataset, 'order_id')

array(['73fc7af87114b39712e6da79b0a377eb',
       'a548910a1c6147796b98fdf73dbeba33',
       'f9e4b658b201a9f2ecdecbb34bed034b', ...,
       '55d4004744368f5571d1f590031933e4',
       '7725825d039fc1f0ceb7635e3f7d9206',
       '90531360ecb1eec2a1fbb265a0db0508'], dtype=object)

In [303]:
sorted(valores_colunas(olist_order_reviews_dataset, 'review_score'))

[1, 2, 3, 4, 5]

In [304]:
valores_colunas(olist_order_reviews_dataset, 'review_comment_title')

array([nan, 'recomendo', 'Super recomendo', ...,
       'Não foi entregue o pedido', 'OTIMA EMBALAGEM', 'Foto enganosa '],
      dtype=object)

In [305]:
valores_colunas(olist_order_reviews_dataset, 'review_comment_message')

array([nan, 'Recebi bem antes do prazo estipulado.',
       'Parabéns lojas lannister adorei comprar pela Internet seguro e prático Parabéns a todos feliz Páscoa',
       ...,
       'O produto não foi enviado com NF, não existe venda sem NF, com certeza fico no aguardo do envio da NF podendo ser por e-mail.',
       'Excelente mochila, entrega super rápida. Super recomendo essa loja!',
       'meu produto chegou e ja tenho que devolver, pois está com defeito , não segurar carga'],
      dtype=object)

In [306]:
valores_colunas(olist_order_reviews_dataset, 'review_creation_date')

array(['2018-01-18 00:00:00', '2018-03-10 00:00:00',
       '2018-02-17 00:00:00', '2017-04-21 00:00:00',
       '2018-03-01 00:00:00', '2018-04-13 00:00:00',
       '2017-07-16 00:00:00', '2018-08-14 00:00:00',
       '2017-05-17 00:00:00', '2018-05-22 00:00:00',
       '2017-12-23 00:00:00', '2017-12-19 00:00:00',
       '2018-02-16 00:00:00', '2018-03-28 00:00:00',
       '2017-04-30 00:00:00', '2018-05-23 00:00:00',
       '2018-03-24 00:00:00', '2017-09-29 00:00:00',
       '2018-08-15 00:00:00', '2017-10-06 00:00:00',
       '2018-03-17 00:00:00', '2018-07-10 00:00:00',
       '2017-09-09 00:00:00', '2017-12-01 00:00:00',
       '2017-06-23 00:00:00', '2017-11-01 00:00:00',
       '2018-03-23 00:00:00', '2018-03-29 00:00:00',
       '2017-12-13 00:00:00', '2017-07-19 00:00:00',
       '2018-04-08 00:00:00', '2018-08-25 00:00:00',
       '2018-06-07 00:00:00', '2017-09-12 00:00:00',
       '2018-06-13 00:00:00', '2018-03-22 00:00:00',
       '2018-08-10 00:00:00', '2018-05-13 00:0

In [307]:
valores_colunas(olist_order_reviews_dataset, 'review_answer_timestamp')

array(['2018-01-18 21:46:59', '2018-03-11 03:05:13',
       '2018-02-18 14:36:24', ..., '2018-03-23 09:10:43',
       '2018-07-02 12:59:13', '2017-07-03 21:01:49'], dtype=object)

### <span style = 'color: #0074D9'> **Alterando tipo de dado das colunas "review_creation_date" e "review_answer_timestamp":** </span>

In [308]:
olist_order_reviews_dataset['review_creation_date'] = pd.to_datetime(olist_order_reviews_dataset['review_creation_date'])
olist_order_reviews_dataset['review_answer_timestamp'] = pd.to_datetime(olist_order_reviews_dataset['review_answer_timestamp'])

olist_order_reviews_dataset[['review_creation_date', 'review_answer_timestamp']].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99224 entries, 0 to 99223
Data columns (total 2 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   review_creation_date     99224 non-null  datetime64[ns]
 1   review_answer_timestamp  99224 non-null  datetime64[ns]
dtypes: datetime64[ns](2)
memory usage: 1.5 MB


In [309]:
olist_order_reviews_dataset.head(5)

review_id                          order_id  \
0  7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1  80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2  228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3  e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4  f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   

   review_score review_comment_title  \
0             4                  NaN   
1             5                  NaN   
2             5                  NaN   
3             5                  NaN   
4             5                  NaN   

                              review_comment_message review_creation_date  \
0                                                NaN           2018-01-18   
1                                                NaN           2018-03-10   
2                                                NaN           2018-02-17   
3              Recebi bem antes do prazo estipulado.           2017-04-21   
4  Parabéns lojas lannister adorei comprar pela I...           2018-03-01   

  review_answer_timestamp  
0     2018-01-18 21:46:59  
1     2018-03-11 03:05:13  
2     2018-02-18 14:36:24  
3     2017-04-21 22:02:06  
4     2018-03-02 10:26:53

### <span style = 'color: #0074D9'> **Contagem de reviews pelos valores da variável "review_score":** </span>

In [310]:
total_de_linhas_review_score = olist_order_reviews_dataset['review_score'].shape[0]
print(f'A variável "revies_score" possui {total_de_linhas_review_score:,} avaliações.')

A variável "revies_score" possui 99,224 avaliações.


In [311]:
contagem_de_notas = olist_order_reviews_dataset[['review_id', 'review_score']].groupby(['review_score']).count()
contagem_de_notas

review_id
review_score           
1                 11424
2                  3151
3                  8179
4                 19142
5                 57328

In [312]:
contagem_de_notas_porcentual = (contagem_de_notas / total_de_linhas_review_score)
contagem_de_notas_porcentual = (contagem_de_notas_porcentual * 100).astype(int).astype(str) + '%'
contagem_de_notas_porcentual

review_id
review_score          
1                  11%
2                   3%
3                   8%
4                  19%
5                  57%

#### <span style = 'color: #cc9900'> *Comentários:* </span>

- Podemos observar que das 99,224 avaliações na variável "review_score", 57% das avaliações possuem nota 5, ou seja, 57,328 são positivas.
- Além disso, as avaliações com notas 2 e 1 representam apenas 14% das avaliações, logo, podemos concluir que a maioria dos produtos vendidos estão dentro ou até mesmo superando a expectativa dos clientes, certo?
- Mas o que leva a ter quase 15,000 avalições negativas (com valores 1 ou 2)? É isso que iremos descobrir nesta análise.

## <span style = 'color: #39CCCC'> **Corrigindo tipo de dados:** </span>

### <span style = 'color: #0074D9'> **Clientes:** </span>

In [313]:
olist_customers_dataset['customer_zip_code_prefix'] = olist_customers_dataset['customer_zip_code_prefix'].apply(str)
olist_customers_dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   customer_id               99441 non-null  object
 1   customer_unique_id        99441 non-null  object
 2   customer_zip_code_prefix  99441 non-null  object
 3   customer_city             99441 non-null  object
 4   customer_state            99441 non-null  object
dtypes: object(5)
memory usage: 3.8+ MB


### <span style = 'color: #0074D9'> **Geolocalização:** </span>

In [314]:
olist_geolocation_dataset['geolocation_zip_code_prefix'] = olist_geolocation_dataset['geolocation_zip_code_prefix'].apply(str)
olist_geolocation_dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000163 entries, 0 to 1000162
Data columns (total 5 columns):
 #   Column                       Non-Null Count    Dtype  
---  ------                       --------------    -----  
 0   geolocation_zip_code_prefix  1000163 non-null  object 
 1   geolocation_lat              1000163 non-null  float64
 2   geolocation_lng              1000163 non-null  float64
 3   geolocation_city             1000163 non-null  object 
 4   geolocation_state            1000163 non-null  object 
dtypes: float64(2), object(3)
memory usage: 38.2+ MB


# <span style = 'color: #3D9970'> **03. Relação entre a variável "review_score":** </span>

## <span style = 'color: #39CCCC'> **Unificando "order_id":** </span>

- Devemos entender a relação das variáveis com nossa variável alvo "review_score". 
- Para que haja um review, deve haver um pedido.
- Por este motivo, será criado um dataset com todas as características ligadas ao "order_id" baseado no dataset "olist_order_items_dataset".

In [315]:
#Unificando as tabelas para que seja possível análisar.
merged_customer_id = olist_orders_dataset.merge(olist_customers_dataset, on = 'customer_id', how = 'left')

merged_order_id = olist_order_items_dataset.merge(olist_order_reviews_dataset, on = 'order_id', how = 'left')
merged_order_id = merged_order_id.merge(olist_products_dataset, on = 'product_id', how = 'left')
merged_order_id = merged_order_id.merge(olist_sellers_dataset, on = 'seller_id', how = 'left')
merged_order_id = merged_order_id.merge(olist_order_payments_dataset, on = 'order_id', how = 'left')
merged_order_id = merged_order_id.merge(merged_customer_id, on = 'order_id', how = 'left')

merged_order_id

order_id  order_item_id  \
0       00010242fe8c5a6d1ba2dd792cb16214              1   
1       00018f77f2f0320c557190d7a144bdd3              1   
2       000229ec398224ef6ca0657da4fc703e              1   
3       00024acbcdf0a6daa1e931b038114c75              1   
4       00042b26cf59d7ce69dfabb4e55b4fd9              1   
...                                  ...            ...   
118305  fffc94f6ce00a00581880bf54a75a037              1   
118306  fffcd46ef2263f404302a634eb57f7eb              1   
118307  fffce4705a9662cd70adb13d4a31832d              1   
118308  fffe18544ffabc95dfada21779c9644f              1   
118309  fffe41c64501cc87c801fd61db3f6244              1   

                              product_id                         seller_id  \
0       4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1       e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2       c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3       7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4       ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   
...                                  ...                               ...   
118305  4aa6014eceb682077f9dc4bffebc05b0  b8bc237ba3788b23da09c0f1f3a3288c   
118306  32e07fd915822b0765e448c4dd74c828  f3c38ab652836d21de61fb8314b69182   
118307  72a30483855e2eafc67aee5dc2560482  c3cfdc648177fdbbbb35635a37472c53   
118308  9c422a519119dcad7575db5af1ba540e  2b3e4a2a3ea8e01938cabda2a3e5cc79   
118309  350688d9dc1e75ff97be326363655e01  f7ccf836d21b2fb1de37564105216cc1   

        shipping_limit_date   price  freight_value  \
0       2017-09-19 09:45:35   58.90          13.29   
1       2017-05-03 11:05:13  239.90          19.93   
2       2018-01-18 14:48:30  199.00          17.87   
3       2018-08-15 10:10:18   12.99          12.79   
4       2017-02-13 13:57:51  199.90          18.14   
...                     ...     ...            ...   
118305  2018-05-02 04:11:01  299.99          43.41   
118306  2018-07-20 04:31:48  350.00          36.53   
118307  2017-10-30 17:14:25   99.90          16.95   
118308  2017-08-21 00:04:32   55.99           8.72   
118309  2018-06-12 17:10:13   43.00          12.79   

                               review_id  review_score review_comment_title  \
0       97ca439bc427b48bc1cd7177abe71365           5.0                  NaN   
1       7b07bacd811c4117b742569b04ce3580           4.0                  NaN   
2       0c5b33dea94867d1ac402749e5438e8b           5.0                  NaN   
3       f4028d019cb58564807486a6aaf33817           4.0                  NaN   
4       940144190dcba6351888cafa43f3a3a5           5.0                  NaN   
...                                  ...           ...                  ...   
118305  9185f849f32d82e216a4e025e0c50f5c           5.0                  NaN   
118306  be803f6a93d64719fd685c1cc610918a           5.0                  NaN   
118307  dbdd81cd59a1a9f94a10a990b4d48dce           5.0                  NaN   
118308  fba117c9ac40d41ca7be54741f471303           5.0                  NaN   
118309  b2700869a37f1aafc9dda829dc2f9027           5.0                  NaN   

        ... order_status order_purchase_timestamp    order_approved_at  \
0       ...    delivered      2017-09-13 08:59:02  2017-09-13 09:45:35   
1       ...    delivered      2017-04-26 10:53:06  2017-04-26 11:05:13   
2       ...    delivered      2018-01-14 14:33:31  2018-01-14 14:48:30   
3       ...    delivered      2018-08-08 10:00:35  2018-08-08 10:10:18   
4       ...    delivered      2017-02-04 13:57:51  2017-02-04 14:10:13   
...     ...          ...                      ...                  ...   
118305  ...    delivered      2018-04-23 13:57:06  2018-04-25 04:11:01   
118306  ...    delivered      2018-07-14 10:26:46  2018-07-17 04:31:48   
118307  ...    delivered      2017-10-23 17:07:56  2017-10-24 17:14:25   
118308  ...    delivered      2017-08-14 23:02:5

In [316]:
# Excluindo order_id duplicados
linhas_duplas = merged_order_id[merged_order_id['order_id'].duplicated()]
merged_order_id.drop_duplicates(subset = 'order_id', keep = 'first', inplace = True)
print(f'Foram excluídos {linhas_duplas.shape[0]} registros da tabela.')

Foram excluídos 19644 registros da tabela.


In [317]:
# Verificando e validando os valores nulos presentes em review_score
valores_nulos_dataset(merged_order_id)

order_id                             0
order_item_id                        0
product_id                           0
seller_id                            0
shipping_limit_date                  0
price                                0
freight_value                        0
review_id                          749
review_score                       749
review_comment_title             87204
review_comment_message           58404
review_creation_date               749
review_answer_timestamp            749
product_category_name             1416
product_name_lenght               1416
product_description_lenght        1416
product_photos_qty                1416
product_weight_g                    16
product_length_cm                   16
product_height_cm                   16
product_width_cm                    16
seller_zip_code_prefix               0
seller_city                          0
seller_state                         0
payment_sequential                   1
payment_type             

#### <span style = 'color: #cc9900'> *Comentários:* </span>

- Podemos observar que realizado a junção das tabelas pelo order_id, visualizamos que nossa variável alvo possui alguns valores em branco (749 registros).
- Para este caso, iremos realizar o preenchimento dos valores nulos utilizando a mediana da variável.

In [318]:
# Transformando os valores nulos na mediana do dataset
mediana_review_score = merged_order_id['review_score'].median()
merged_order_id['review_score'] = merged_order_id['review_score'].fillna(mediana_review_score)

print(f'''Após o procedimento acima, podemos visualizar que há um total de {merged_order_id['review_score'].isnull().sum()} registros nulos 
e há {merged_order_id[merged_order_id['order_id'].duplicated()].shape[0]} registros duplicados.''')

Após o procedimento acima, podemos visualizar que há um total de 0 registros nulos 
e há 0 registros duplicados.


## <span style = 'color: #39CCCC'> **Possíveis relações:** </span>

[  ] Review com **cidade da venda**. <br>
[  ] Review com **tempo de entrega**. <br>

- Realizada a união dos datasets, podemos trabalhar com nossas variáveis e identificar qual a relação entre as variáveis com nossa variável alvo.

### <span style = 'color: #0074D9'> **Cidade da venda:** </span>